src/System.Buffers/src/System/Buffers/ArrayPool.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// This class provides resource pooling for arrays of any type. This is can increase performance
+    /// of certain applications where lots of arrays are created and destroyed in rapid succession.
+    /// This class is thread-safe.
+    /// </summary>
+    public abstract class ArrayPool<T>
+    {
+        /// <summary>The default number of arrays that are available for rent.</summary>
+        public const int DefaultNumberOfArraysPerBucket = 50;
+        /// <summary>The default length of each Rent'able array; equal to 1MB in bytes</summary>
+        public const int DefaultArrayLength = 1024000;
+
+        private static ArrayPool<T> _sharedInstance = null;
+
+        /// <summary>
+        /// Retrieves a shared instance pool that is thread-safe and can be used by 
+        /// multiple components of the same system.
+        /// </summary>
+        public static ArrayPool<T> Shared
+        {
+            get
+            {
+                if (Volatile.Read(ref _sharedInstance) == null)
+                {
+                    Interlocked.CompareExchange(ref _sharedInstance, new DefaultArrayPool<T>(DefaultArrayLength, DefaultNumberOfArraysPerBucket), null);
+                }
+
+                return _sharedInstance;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new ArrayPool instance of the given type.
+        /// </summary>
+        /// <param name="maxArrayLength">The maximum length of any Rent request.</param>
+        /// <param name-"numberOfArrays">The maximum number of arrays that will be rented.</param>
+        /// <returns>Returns a new ArrayPool<T> instance with the specified configuration options</return>
+        public static ArrayPool<T> Create(int maxArrayLength = DefaultArrayLength, int numberOfArrays = DefaultNumberOfArraysPerBucket)
+        {
+            return new DefaultArrayPool<T>(maxArrayLength, numberOfArrays);
+        }
+
+        /// <summary>
+        /// Gets the number of buffers that have been Rented and not yet Returned. This count includes
+        /// any buffers that have been allocated on-demand.
+        /// </summary>
+        public abstract int RentedBufferCount { get; }
+
+        /// <summary>
+        /// Gets the total number of allocated buffers available for Rent. This count includes any outstanding
+        /// buffers that have been allocated on-demand.
+        /// </summary>
+        public abstract int TotalBufferCount { get; }
+
+        /// <summary>
+        /// Retrieves a buffer from the pool that is at least the requested length. This buffer is loaned
+        /// to the caller and the caller must call Return when finished with the buffer.
+        /// </summary>
```

Comment:
> once we figure out the "overriding the shared pool creation" issue

I didn't realize we were considering that.  We need to be very careful about the kind of customization allowed there, as components (including some of ours low in the stack) may take a dependency on aspects of the policy.  For example, if System.IO.Compression uses ```ArrayPool<byte>.Shared.Rent(8192)``` to get its byte[], Rent better not start returning null or throwing if no buffer is available.

> the need for async APIs or TryRent goes away

Can you elaborate on how?  For example, let's say I want my own type like this (pseudo-code):
```C#
internal sealed class AsyncArrayPool<T>
{
    private readonly Queue<TaskCompletionSource<T[]>> _waiters = new Queue<TaskCompletionSource<T[]>>();
    private readonly int _length;

    public AsyncArrayPool(int length) { _length = length; }

    public ValueTask<T[]> RentAsync()
    {
        lock (_waiters)
        {
            T[] buffer;
            if (ArrayPool<byte>.Shared.TryRent(_length, out buffer))
                return buffer;

            var tcs = new TaskCompletionSource<T[]>(TaskCreationOptions.RunContinuationsAsynchronously);
            _waiters.Enqueue(tcs);
            return tcs.Task;
        }
    }

    public void Return(T[] buffer)
    {
        lock (_waiters)
        {
            if (_waiters.Count > 0)
            {
                 var waiter = _waiters.Dequeue();
                 waiter.SetResult(buffer);
            }   
            else ArrayPool<byte>.Shared.Return(buffer);
        }
    }
}
```
I'm not sure how I do that without something like TryRent (or Rent return null / throwing, but I already stated I don't think that's a policy that can be substituted).